File: solutions/456.132-pattern/132-pattern.cpp
class Solution {
public:
    bool find132pattern(vector<int>& nums) {
        int n = nums.size(), top = n, third = INT_MIN;

        for (int i = n - 1; i >= 0; i--) {
            if (nums[i] < third) return true;
            while (top < n && nums[i] > nums[top]) third = nums[top++];
            nums[--top] = nums[i];
        }

        return false;
    }
};

File: solutions/456.132-pattern/132-pattern.java
class Solution {
    public boolean find132pattern(int[] nums) {
        Stack<Integer> stack = new Stack<>();
        int ak = Integer.MIN_VALUE;
        for(int i = nums.length - 1; i >= 0; i--) {
            if(nums[i] < ak) return true;
            while(!stack.isEmpty()) {
                if(nums[i] > stack.peek()) {
                    ak = stack.pop();
                } else break;
            }
            stack.push(nums[i]);
        }
        return false;
    }
}